chain/types/ethtypes/eth_transactions.go

```diff
@@ -227,347 +184,90 @@ func (tx *EthTxArgs) ToSignedMessage() (*types.SignedMessage, error) {
 	}, nil
 }
 
-func (tx *EthTxArgs) HashedOriginalRlpMsg() ([]byte, error) {
-	msg, err := tx.ToRlpUnsignedMsg()
-	if err != nil {
-		return nil, err
-	}
-
-	hasher := sha3.NewLegacyKeccak256()
-	hasher.Write(msg)
-	hash := hasher.Sum(nil)
-	return hash, nil
-}
-
-func (tx *EthTxArgs) ToRlpUnsignedMsg() ([]byte, error) {
-	packed, err := tx.packTxFields()
-	if err != nil {
-		return nil, err
-	}
-
-	encoded, err := EncodeRLP(packed)
-	if err != nil {
-		return nil, err
-	}
-	return append([]byte{0x02}, encoded...), nil
-}
-
-func (tx *EthTx) ToEthTxArgs() EthTxArgs {
-	return EthTxArgs{
-		ChainID:              int(tx.ChainID),
-		Nonce:                int(tx.Nonce),
-		To:                   tx.To,
-		Value:                big.Int(tx.Value),
-		MaxFeePerGas:         big.Int(tx.MaxFeePerGas),
-		MaxPriorityFeePerGas: big.Int(tx.MaxPriorityFeePerGas),
-		GasLimit:             int(tx.Gas),
-		Input:                tx.Input,
-		V:                    big.Int(tx.V),
-		R:                    big.Int(tx.R),
-		S:                    big.Int(tx.S),
+func ParseEthTransaction(data []byte) (EthTransaction, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("empty data")
 	}
-}
-
-func (tx *EthTx) TxHash() (EthHash, error) {
-	ethTxArgs := tx.ToEthTxArgs()
-	return (&ethTxArgs).TxHash()
-}
 
-func (tx *EthTxArgs) TxHash() (EthHash, error) {
-	rlp, err := tx.ToRlpSignedMsg()
-	if err != nil {
-		return EmptyEthHash, err
+	if data[0] > 0x7f {
+		tx, err := parseLegacyHomesteadTx(data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse legacy homestead transaction: %w", err)
+		}
+		return tx, nil
 	}
 
-	return EthHashFromTxBytes(rlp), nil
-}
-
-func (tx *EthTxArgs) ToRlpSignedMsg() ([]byte, error) {
-	packed1, err := tx.packTxFields()
-	if err != nil {
-		return nil, err
+	if data[0] == 1 {
```

Comment:
looks like this whole section could be a `switch`

Reply:
Done.